### ISSUE TITLE
Returned generated ID on INSERT is incorrect when using ON CONFLICT IGNORE clauses

### ISSUE THREAD

**Issue by recastrodiaz** [tekartik/sqflite#164]
When inserting a row with conflicts and conflicts are set to IGNORE, the returned id should be null or 0 as the row with this id does not exist or points to a different row.

### Database structure:

```SQL
CREATE TABLE test (
	id INTEGER PRIMARY KEY AUTOINCREMENT,
	label TEXT NOT NULL,
	UNIQUE (label) ON CONFLICT IGNORE
);
```

### How to replicate

```DART
int id = await database.rawInsert('''
        INSERT INTO test (label) VALUES(?)
        ''', [
      'label-1'
    ]);
print(id); // prints 1

id = await database.rawInsert('''
        INSERT INTO test (label) VALUES(?)
        ''', [
      'label-2'
    ]);
print(id); // prints 2

id = await database.rawInsert('''
        INSERT INTO test (label) VALUES(?)
        ''', [
      'label-1'
    ]);
print(id); // prints 2 instead of 1 or null or 0

```

### Root cause

The issue on Android is that the Java code assumes that `last_insert_rowid()` always returns the last inserted row id. However, when INSERT OR REPLACE encounters a constraint violation and the resolution is ON CONFLICT IGNORE, no row is inserted and no exception is thrown, yet a non null row id is returned.

TODO: I have yet to test this on iOS

https://www.sqlite.org/c3ref/last_insert_rowid.html

> An INSERT that fails due to a constraint violation is not a successful INSERT and does not change the value returned by this routine. Thus INSERT OR FAIL, INSERT OR IGNORE, INSERT OR ROLLBACK, and INSERT OR ABORT make no changes to the return value of this routine when their insertion fails. When INSERT OR REPLACE encounters a constraint violation, it does not fail. The INSERT continues to completion after deleting rows that caused the constraint problem so INSERT OR REPLACE will always change the return value of this interface.

### Possible solution

Use `changes()` (when > 0) coupled with `last_insert_rowid()` to decide whether to return `last_insert_rowid` or null in https://github.com/tekartik/sqflite/blob/de4b89b832822c1a5a54681d661d5ffc8064badc/android/src/main/java/com/tekartik/sqflite/SqflitePlugin.java#L420:

```SQL
INSERT INTO test(label) VALUES ("label-1");
SELECT last_insert_rowid(), changes(); -- Returns: 1, 1

INSERT INTO test(label) VALUES ("label-2");
SELECT last_insert_rowid(), changes(); -- Returns: 2, 1

INSERT INTO test(label) VALUES ("label-1");
SELECT last_insert_rowid(), changes(); -- Returns: 2, 0
```

**Comment by alextekartik** [comment]
Thanks for the report, the detailed explanation, ways to reproduce it and possible solution. I will look into it

**Comment by recastrodiaz** [comment]
Thanks for the quick response and BTW superb work with the library! It's super useful and a big time saver!

**Comment by alextekartik** [comment]
Your solution was a good one. I had to do it on both iOS and Android. If the number of changes is 0, it now properly returns null. It adds an extra sql call for each insert.
Published in 1.1.0+1. Let me know if that works on your side

Thanks again for your detailed report. Reproducing it (I used your code for the test) and fixing it was straightforward!

**Comment by recastrodiaz** [comment]
Wow! That was a super quick fix. Thanks!
I confirm 1.1.0+1 works as expected my side. Thanks again!
